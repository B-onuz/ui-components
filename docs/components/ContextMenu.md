# ContextMenu (Component)

[Descrição do componente e onde usar]

## Como utilizar

```js
import ContextMenu from 'ui-components/lib/components/ContextMenu'
```

```jsx
import React from 'react'
import ContextMenu from 'ui-components/lib/components/ContextMenu'
import { Flex, Box } from 'reflexbox/styled-components'

const functionA = () => {
  console.log('Função A')
}

const functionB = () => {
  console.log('Função B')
}

const functionC = () => {
  console.log('Função C')
}

const exampleObj = [
  {
    text: 'editar',
    buttonFunction: functionA,
  },
  {
    text: 'adicionar',
    buttonFunction: functionB,
  },
  {
    text: 'excluir',
    buttonFunction: functionC,
  },
]

export const ContextMenuExample = () => (
  <Flex style={{ height: '100vh' }} width={[1]} justifyContent={'space-evenly'} alignItems={'center'}>
    <Box p={4}>
      <ContextMenu contextMenuActions={exampleObj} direction={'right'} />
    </Box>
    <Box p={4}>
      <ContextMenu contextMenuActions={exampleObj} direction={'top'} />
    </Box>
    <Box p={4}>
      <ContextMenu contextMenuActions={exampleObj} direction={'left'} />
    </Box>
    <Box p={4}>
      <ContextMenu contextMenuActions={exampleObj} direction={'bottom'} />
    </Box>
  </Flex>
)
```

Escala no tema: `theme.variantScale`

## Styled-system props

- space https://styled-system.com/table#space
- layout https://styled-system.com/table#layout
