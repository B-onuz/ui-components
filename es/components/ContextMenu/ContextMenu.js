"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _styledSystem = require("styled-system");

var _reactFontawesome = require("@fortawesome/react-fontawesome");

var _freeSolidSvgIcons = require("@fortawesome/free-solid-svg-icons");

var _Button = _interopRequireDefault(require("../Button"));

var _this = void 0,
    _jsxFileName = "/Users/rogeralbino/projects/Kazap/B-onuz/ui-components/src/components/ContextMenu/ContextMenu.js";

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _templateObject5() {
  var data = _taggedTemplateLiteralLoose(["\n  color: ", ";\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = _taggedTemplateLiteralLoose(["\n  width: 100%;\n  margin: 8px 0;\n  border: 1px solid #f1f1f3;\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteralLoose(["\n  position: absolute;\n  border-radius: 12px;\n  opacity: ", ";\n  z-index: ", ";\n  display: flex;\n  flex-direction: column;\n  align-items: center;\n  padding: 8px 16px;\n  background: #fff 0% 0% no-repeat padding-box;\n  box-shadow: 0px 2px 6px #2c28281c;\n  transition: all 300ms;\n\n  ", "\n  &::after {\n    content: '';\n    width: 1em;\n    height: 1em;\n    position: absolute;\n    box-shadow: 1px 2px 2px -1px #2c28281c;\n    background: #fff 0% 0%;\n    ", "\n  }\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteralLoose(["\n  width: 2em;\n  height: 2em;\n  border-radius: 50%;\n  box-shadow: 0px 2px 3px #2c282848;\n  border: none;\n  &:hover {\n    cursor: pointer;\n  }\n  &:focus {\n    outline: none;\n  }\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["\n  ", "\n  ", "\n  position: relative;\n  display: flex;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteralLoose(strings, raw) { if (!raw) { raw = strings.slice(0); } strings.raw = raw; return strings; }

var StyledContextMenu = _styledComponents["default"].div(_templateObject(), _styledSystem.color, _styledSystem.space);

var ContextMenuToogleButton = _styledComponents["default"].button(_templateObject2());

var ContextDialogMenu = _styledComponents["default"].div(_templateObject3(), function (_ref) {
  var isOpen = _ref.isOpen;
  return isOpen ? 1 : 0;
}, function (_ref2) {
  var isOpen = _ref2.isOpen;
  return isOpen ? 1 : -100;
}, function (_ref3) {
  var direction = _ref3.direction;

  if (direction === 'left') {
    return "\n      right: calc(100% + 24px);\n      top: -50%;\n    ";
  }

  if (direction === 'right') {
    return "\n      left: calc(100% + 24px);\n      top: -50%;\n    ";
  }

  if (direction === 'top') {
    return "\n      left: 50%;\n      transform: translateX(-50%);\n      bottom: calc(100% + 1.5em);\n    ";
  }

  if (direction === 'bottom') {
    return "\n      transform: translateX(-50%);\n      left: 50%;\n      top: calc(100% + 1.5em);\n    ";
  }
}, function (_ref4) {
  var direction = _ref4.direction;

  if (direction === 'left') {
    return "\n      right: -0.5em;\n      top: 1.6em;\n      transform: rotateZ(-45deg);\n    ";
  }

  if (direction === 'right') {
    return "\n      left: -0.5em;\n      top: 1.6em;\n      transform: rotateZ(135deg);\n    ";
  }

  if (direction === 'top') {
    return "\n      top: calc(100% - 0.5em);\n      transform: rotateZ(45deg);\n      left: 0;\n      right: 0;\n      margin: auto;\n    ";
  }

  if (direction === 'bottom') {
    return "\n      bottom: calc(100% - 0.5em);\n      transform: rotateZ(-135deg);\n      left: 0;\n      right: 0;\n      margin: auto;\n    ";
  }
});

var Divider = _styledComponents["default"].hr(_templateObject4());

var Icon = (0, _styledComponents["default"])(_reactFontawesome.FontAwesomeIcon)(_templateObject5(), function (_ref5) {
  var theme = _ref5.theme;
  return theme.colors.darkGrey;
});

var ContextMenu = function ContextMenu(_ref6) {
  var children = _ref6.children,
      direction = _ref6.direction,
      contextMenuActions = _ref6.contextMenuActions,
      rest = _objectWithoutPropertiesLoose(_ref6, ["children", "direction", "contextMenuActions"]);

  var _useState = (0, _react.useState)(false),
      isOpen = _useState[0],
      setIsOpen = _useState[1];

  var wrapperRef = (0, _react.useRef)(null);

  var useOutsideAlerter = function useOutsideAlerter(ref) {
    (0, _react.useEffect)(function () {
      function handleClickOutside(event) {
        if (ref.current && !ref.current.contains(event.target)) {
          setIsOpen(false);
        }
      }

      document.addEventListener('mousedown', handleClickOutside);
      return function () {
        document.removeEventListener('mousedown', handleClickOutside);
      };
    }, [ref]);
  };

  useOutsideAlerter(wrapperRef);
  return /*#__PURE__*/_react["default"].createElement(StyledContextMenu, _extends({
    ref: wrapperRef
  }, rest, {
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 146,
      columnNumber: 5
    }
  }), /*#__PURE__*/_react["default"].createElement(ContextMenuToogleButton, {
    onClick: function onClick() {
      return setIsOpen(!isOpen);
    },
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 147,
      columnNumber: 7
    }
  }, /*#__PURE__*/_react["default"].createElement(Icon, {
    icon: _freeSolidSvgIcons.faEllipsisH,
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 148,
      columnNumber: 9
    }
  })), /*#__PURE__*/_react["default"].createElement(ContextDialogMenu, {
    isOpen: isOpen,
    direction: direction,
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 150,
      columnNumber: 7
    }
  }, (contextMenuActions || []).map(function (action, index) {
    return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement(_Button["default"], {
      linkButton: true,
      color: 'primary',
      onClick: function onClick() {
        return action.buttonFunction();
      },
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 153,
        columnNumber: 13
      }
    }, action.text), index + 1 !== contextMenuActions.length ? /*#__PURE__*/_react["default"].createElement(Divider, {
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 156,
        columnNumber: 56
      }
    }) : null);
  })));
};

ContextMenu.defaultProps = {
  direction: 'left'
};
var _default = ContextMenu;
exports["default"] = _default;