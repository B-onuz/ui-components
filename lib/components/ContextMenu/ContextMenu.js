"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _styledSystem = require("styled-system");

var _reactFontawesome = require("@fortawesome/react-fontawesome");

var _freeSolidSvgIcons = require("@fortawesome/free-solid-svg-icons");

var _Button = _interopRequireDefault(require("../Button"));

var _ContextMenuDialog = _interopRequireDefault(require("./ContextMenuDialog"));

var _this = void 0,
    _jsxFileName = "/home/gabriel/Documentos/Projetos/Bonuz/ui-components/src/components/ContextMenu/ContextMenu.js";

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _templateObject4() {
  var data = _taggedTemplateLiteralLoose(["\n  color: ", ";\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteralLoose(["\n  width: 100%;\n  margin: 8px 0;\n  border: 1px solid #f1f1f3;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteralLoose(["\n  width: 2em;\n  height: 2em;\n  border-radius: 50%;\n  box-shadow: 0px 2px 3px #2c282848;\n  border: none;\n  &:hover {\n    cursor: pointer;\n  }\n  &:focus {\n    outline: none;\n  }\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["\n  ", "\n  ", "\n  position: relative;\n  display: flex;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteralLoose(strings, raw) { if (!raw) { raw = strings.slice(0); } strings.raw = raw; return strings; }

var StyledContextMenu = _styledComponents["default"].div(_templateObject(), _styledSystem.color, _styledSystem.space);

var ContextMenuToogleButton = _styledComponents["default"].button(_templateObject2());

var Divider = _styledComponents["default"].hr(_templateObject3());

var Icon = (0, _styledComponents["default"])(_reactFontawesome.FontAwesomeIcon)(_templateObject4(), function (_ref) {
  var theme = _ref.theme;
  return theme.colors.darkGrey;
});

var ContextMenu = function ContextMenu(_ref2) {
  var children = _ref2.children,
      direction = _ref2.direction,
      contextMenuActions = _ref2.contextMenuActions,
      rest = _objectWithoutPropertiesLoose(_ref2, ["children", "direction", "contextMenuActions"]);

  var _useState = (0, _react.useState)(false),
      isOpen = _useState[0],
      setIsOpen = _useState[1];

  var wrapperRef = (0, _react.useRef)(null);
  var buttonRef = (0, _react.useRef)(null);

  var onClose = function onClose() {
    setIsOpen(false);
    return true;
  };

  return /*#__PURE__*/_react["default"].createElement(StyledContextMenu, _extends({
    ref: wrapperRef
  }, rest, {
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 51,
      columnNumber: 5
    }
  }), /*#__PURE__*/_react["default"].createElement(ContextMenuToogleButton, {
    ref: buttonRef,
    onClick: function onClick() {
      return setIsOpen(function (last) {
        return !last;
      });
    },
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 52,
      columnNumber: 7
    }
  }, /*#__PURE__*/_react["default"].createElement(Icon, {
    icon: _freeSolidSvgIcons.faEllipsisH,
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 53,
      columnNumber: 9
    }
  })), /*#__PURE__*/_react["default"].createElement(_ContextMenuDialog["default"], {
    isOpen: isOpen,
    onClose: onClose,
    buttonRef: buttonRef,
    direction: direction,
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 55,
      columnNumber: 7
    }
  }, (contextMenuActions || []).map(function (action, index) {
    return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement(_Button["default"], {
      linkButton: true,
      color: 'primary',
      onClick: function onClick() {
        return onClose() && action.buttonFunction();
      },
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 58,
        columnNumber: 13
      }
    }, action.text), index + 1 !== contextMenuActions.length ? /*#__PURE__*/_react["default"].createElement(Divider, {
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 61,
        columnNumber: 56
      }
    }) : null);
  })));
};

ContextMenu.defaultProps = {
  direction: 'left'
};
var _default = ContextMenu;
exports["default"] = _default;
module.exports = exports.default;