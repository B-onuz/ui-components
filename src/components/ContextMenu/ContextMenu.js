import React, { useRef, useState, useEffect } from 'react'
import styled from 'styled-components'
import { color, space } from 'styled-system'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faEllipsisH } from '@fortawesome/free-solid-svg-icons'
import Button from '../Button'

const StyledContextMenu = styled.div`
  ${color}
  ${space}
  position: relative;
  display: flex;
`

const ContextMenuToogleButton = styled.button`
  width: 2em;
  height: 2em;
  border-radius: 50%;
  box-shadow: 0px 2px 3px #2c282848;
  border: none;
  &:hover {
    cursor: pointer;
  }
  &:focus {
    outline: none;
  }
`

const ContextDialogMenu = styled.div`
  position: absolute;
  border-radius: 12px;
  opacity: ${({ isOpen }) => (isOpen ? 1 : 0)};
  z-index: ${({ isOpen }) => (isOpen ? 1 : -100)};
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 8px 16px;
  background: #fff 0% 0% no-repeat padding-box;
  box-shadow: 0px 2px 6px #2c28281c;
  transition: all 300ms;

  ${({ direction }) => {
    if (direction === 'left') {
      return `
      right: calc(100% + 24px);
      top: -50%;
    `
    }
    if (direction === 'right') {
      return `
      left: calc(100% + 24px);
      top: -50%;
    `
    }
    if (direction === 'top') {
      return `
      left: 50%;
      transform: translateX(-50%);
      bottom: calc(100% + 1.5em);
    `
    }
    if (direction === 'bottom') {
      return `
      transform: translateX(-50%);
      left: 50%;
      top: calc(100% + 1.5em);
    `
    }
  }}
  &::after {
    content: '';
    width: 1em;
    height: 1em;
    position: absolute;
    box-shadow: 1px 2px 2px -1px #2c28281c;
    background: #fff 0% 0%;
    ${({ direction }) => {
      if (direction === 'left') {
        return `
      right: -0.5em;
      top: 1.6em;
      transform: rotateZ(-45deg);
    `
      }
      if (direction === 'right') {
        return `
      left: -0.5em;
      top: 1.6em;
      transform: rotateZ(135deg);
    `
      }
      if (direction === 'top') {
        return `
      top: calc(100% - 0.5em);
      transform: rotateZ(45deg);
      left: 0;
      right: 0;
      margin: auto;
    `
      }
      if (direction === 'bottom') {
        return `
      bottom: calc(100% - 0.5em);
      transform: rotateZ(-135deg);
      left: 0;
      right: 0;
      margin: auto;
    `
      }
    }}
  }
`

const Divider = styled.hr`
  width: 100%;
  margin: 8px 0;
  border: 1px solid #f1f1f3;
`

const Icon = styled(FontAwesomeIcon)`
  color: ${({ theme }) => theme.colors.darkGrey};
`

const ContextMenu = ({ children, direction, contextMenuActions, ...rest }) => {
  const [isOpen, setIsOpen] = useState(false)
  const wrapperRef = useRef(null)

  const useOutsideAlerter = (ref) => {
    useEffect(() => {
      function handleClickOutside(event) {
        if (ref.current && !ref.current.contains(event.target)) {
          setIsOpen(false)
        }
      }

      document.addEventListener('mousedown', handleClickOutside)
      return () => {
        document.removeEventListener('mousedown', handleClickOutside)
      }
    }, [ref])
  }

  useOutsideAlerter(wrapperRef)

  return (
    <StyledContextMenu ref={wrapperRef} {...rest}>
      <ContextMenuToogleButton onClick={() => setIsOpen(!isOpen)}>
        <Icon icon={faEllipsisH} />
      </ContextMenuToogleButton>
      <ContextDialogMenu isOpen={isOpen} direction={direction}>
        {(contextMenuActions || []).map((action, index) => (
          <>
            <Button linkButton color={'primary'} onClick={() => action.buttonFunction()}>
              {action.text}
            </Button>
            {index + 1 !== contextMenuActions.length ? <Divider /> : null}
          </>
        ))}
      </ContextDialogMenu>
    </StyledContextMenu>
  )
}

ContextMenu.defaultProps = {
  direction: 'left',
}

export default ContextMenu
