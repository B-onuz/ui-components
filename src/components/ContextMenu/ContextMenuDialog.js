import React, { useState, useEffect } from 'react'
import styled from 'styled-components'
import ReactDOM from 'react-dom'

const Dialog = styled.div`
  position: absolute;
  border-radius: 12px;
  opacity: ${({ isOpen }) => (isOpen ? 1 : 0)};
  z-index: ${({ isOpen }) => (isOpen ? 1 : -100)};
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 10px 16px;
  background: #fff 0% 0% no-repeat padding-box;
  box-shadow: 0px 2px 6px #2c28281c;

  ${({ direction, top, right, left, bottom }) => {
    if (direction === 'left') {
      return `
      top: ${top};
      left: ${left};
      transform: translateX(calc(-100% - 32px * 2  - 5px));
    `
    }
    if (direction === 'right') {
      return `
      top: ${top};
      left: ${left};
    `
    }
    if (direction === 'top') {
      return `
      top: ${top};
      left: ${left};
      transform: translateX(calc(-50% - 32px / 2)) translateY(calc(-100% - 24px));
    `
    }
    if (direction === 'bottom') {
      return `
      top: ${top};
      left: ${left};
      transform: translateX(calc(-50% - 32px / 2));
    `
    }
  }}
  &::after {
    content: '';
    width: 1em;
    height: 1em;
    position: absolute;
    box-shadow: 1px 2px 2px -1px #2c28281c;
    background: #fff 0% 0%;
    ${({ direction }) => {
      if (direction === 'left') {
        return `
      right: -0.5em;
      top: 1.6em;
      transform: rotateZ(-45deg);
    `
      }
      if (direction === 'right') {
        return `
      left: -0.5em;
      top: 1.6em;
      transform: rotateZ(135deg);
    `
      }
      if (direction === 'top') {
        return `
      top: calc(100% - 0.5em);
      transform: rotateZ(45deg);
      left: 0;
      right: 0;
      margin: auto;
    `
      }
      if (direction === 'bottom') {
        return `
      bottom: calc(100% - 0.5em);
      transform: rotateZ(-135deg);
      left: 0;
      right: 0;
      margin: auto;
    `
      }
    }}
  }

  @media only screen and (max-width: 500px) {
    position: fixed;
    margin: 0;
    top: 50%;
    left: 50%;
    margin-right: -50%;
    transform: translate(-50%, -50%);
    padding: 20px 30px;

    &::after {
      display: none;
    }
  }
`

const useRootDialog = () => {
  const [$rootDialog, setRootDialog] = useState()
  const IdName = 'contextMenu-root'

  useEffect(() => {
    const rootModal = document.getElementById(IdName)
    if (rootModal) {
      setRootDialog(rootModal)
    } else {
      const rootEl = document.createElement('div')
      rootEl.id = IdName
      setRootDialog(rootEl)
      document.body.appendChild(rootEl)
    }
  })

  return { $rootDialog }
}

const Wrapper = styled.div`
  z-index: 99999999999;
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
`

const ContextMenuDialog = ({ isOpen, buttonRef, onClose, direction, contextFunctions, ...rest }) => {
  const { $rootDialog } = useRootDialog()
  const [positions, setPositions] = useState({ top: 0, right: 0, bottom: 0, left: 0 })
  const $currentButtonRef = (buttonRef || {}).current
  const [top, setTop] = useState(0)
  const [left, setLeft] = useState(0)

  const updatePositions = () => {
    if (!$currentButtonRef) return null
    const offset = $currentButtonRef.getBoundingClientRect()

    if (direction === 'top') {
      setTop(`${offset.top + offset.height - 24}px`)
      setLeft(`${offset.left + offset.width}px`)
    }

    if (direction === 'right') {
      setTop(`${offset.top - offset.height / 2}px`)
      setLeft(`${offset.left + offset.width + 20}px`)
    }

    if (direction === 'left') {
      setTop(`${offset.top - offset.height / 2}px`)
      setLeft(`${offset.left + offset.width + 20}px`)
    }

    if (direction === 'bottom') {
      setTop(`${offset.top + offset.height + 24}px`)
      setLeft(`${offset.left + offset.width}px`)
    }
  }

  useEffect(() => {
    updatePositions()
    window.addEventListener('resize', updatePositions)
    return () => {
      window.removeEventListener('resize', updatePositions)
    }
  })

  if (!$rootDialog) return null

  if (!isOpen) {
    return null
  }

  return ReactDOM.createPortal(
    <Wrapper onClick={onClose}>
      <Dialog direction={direction} isOpen={isOpen} top={top} left={left} {...rest} contextFunctions={contextFunctions} />
    </Wrapper>,
    $rootDialog
  )
}

export default ContextMenuDialog
