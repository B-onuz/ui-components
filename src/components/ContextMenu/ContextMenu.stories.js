import React from 'react'
import ContextMenu from './ContextMenu'
import { Flex, Box } from 'reflexbox/styled-components'
import { action } from '@storybook/addon-actions'

export default {
  title: 'ContextMenu',
  component: ContextMenu,
}

const functionA = action('Função A')

const functionB = action('Função B')

const functionC = action('Função C')

const exampleObj = [
  {
    text: 'editar',
    buttonFunction: functionA,
  },
  {
    text: 'adicionar',
    buttonFunction: functionB,
  },
  {
    text: 'excluir',
    buttonFunction: functionC,
  },
]

export const ContextMenuExample = () => (
  <Flex style={{ height: '100vh' }} width={[1]} justifyContent={'space-evenly'} alignItems={'center'}>
    <Box p={4}>
      <ContextMenu contextMenuActions={exampleObj} direction={'right'} />
    </Box>
    <Box p={4}>
      <ContextMenu contextMenuActions={exampleObj} direction={'top'} />
    </Box>
    <Box p={4}>
      <ContextMenu contextMenuActions={exampleObj} direction={'left'} />
    </Box>
    <Box p={4}>
      <ContextMenu contextMenuActions={exampleObj} direction={'bottom'} />
    </Box>
  </Flex>
)
