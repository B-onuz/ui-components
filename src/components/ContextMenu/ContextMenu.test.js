import React from 'react'
import ContextMenu from './ContextMenu.js'
import renderer from 'react-test-renderer'

test('ContextMenu changes the class when hovered', () => {
  const component = renderer.create(<ContextMenu />)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
