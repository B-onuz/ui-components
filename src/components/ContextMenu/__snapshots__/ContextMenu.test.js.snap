// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ContextMenu 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  width: 2em;
  height: 2em;
  border-radius: 50%;
  box-shadow: 0px 2px 3px #2c282848;
  border: none;
}

.c1:hover {
  cursor: pointer;
}

.c1:focus {
  outline: none;
}

.c3 {
  position: absolute;
  border-radius: 12px;
  opacity: 0;
  z-index: -100;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px 16px;
  background: #fff 0% 0% no-repeat padding-box;
  box-shadow: 0px 2px 6px #2c28281c;
  -webkit-transition: all 300ms;
  transition: all 300ms;
  right: calc(100% + 24px);
  top: -50%;
}

.c3::after {
  content: '';
  width: 1em;
  height: 1em;
  position: absolute;
  box-shadow: 1px 2px 2px -1px #2c28281c;
  background: #fff 0% 0%;
  right: -0.5em;
  top: 1.6em;
  -webkit-transform: rotateZ(-45deg);
  -ms-transform: rotateZ(-45deg);
  transform: rotateZ(-45deg);
}

.c2 {
  color: #9A9A9A;
}

<div
  className="c0"
>
  <button
    className="c1"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-ellipsis-h fa-w-16 c2"
      data-icon="ellipsis-h"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M328 256c0 39.8-32.2 72-72 72s-72-32.2-72-72 32.2-72 72-72 72 32.2 72 72zm104-72c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72zm-352 0c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </button>
  <div
    className="c3"
    direction="left"
  />
</div>
`;
